### PR TITLE
Modify headpose Rect.

### PR DESCRIPTION
**Pull request by acured**
Please use `[API Name] Brief description` as title.

## API Name (Kind)
`Face`, `TextAnalytics`, `ComputerVision`...

## Purpose
Describe the intention of the changes being proposed.
What problem does it solve or functionality does it add?

## Other Information
Add any other helpful information that may be needed here.

### CODE REVIEW

File: app-samples/Cognitive-Services-Face-WPF/Sample-WPF/Controls/Face.cs [hunk 1]
```diff
@@ -327,6 +376,23 @@ public int Width
             }
         }
 
+        /// <summary>
+        /// Gets or sets face Angle
```
Comment (lebronJ) [inline app-samples/Cognitive-Services-Face-WPF/Sample-WPF/Controls/Face.cs]: [Nit] `angle`

File: app-samples/Cognitive-Services-Face-WPF/Sample-WPF/Controls/FaceDetectionPage.xaml.cs [hunk 2]
```diff
@@ -79,11 +79,21 @@ public partial class FaceDetectionPage : Page, INotifyPropertyChanged
         /// </summary>
         private ObservableCollection<Face> _resultCollection = new ObservableCollection<Face>();
 
+        /// <summary>
+        /// Headpose results container
```
Comment (lebronJ) [inline app-samples/Cognitive-Services-Face-WPF/Sample-WPF/Controls/FaceDetectionPage.xaml.cs]: [Nit] `Container of face detection results with faces rotated by head poses`

File: app-samples/Cognitive-Services-Face-WPF/Sample-WPF/Controls/FaceDetectionPage.xaml.cs [hunk 3]
```diff
@@ -79,11 +79,21 @@ public partial class FaceDetectionPage : Page, INotifyPropertyChanged
...
+        /// </summary>
+        private ObservableCollection<Face> _resultCollectionWithHeadpose = new ObservableCollection<Face>();
+
         /// <summary>
         /// Image used for rendering and detecting
         /// </summary>
         private ImageSource _selectedFile;
 
+        /// <summary>
+        /// Whether draw headpose
```
Comment (lebronJ) [inline app-samples/Cognitive-Services-Face-WPF/Sample-WPF/Controls/FaceDetectionPage.xaml.cs]: `Whether to draw rectangles rotated by head poses`

File: app-samples/Cognitive-Services-Face-WPF/Sample-WPF/Controls/FaceDetectionPage.xaml.cs [hunk 4]
```diff
@@ -198,6 +208,42 @@ public ImageSource SelectedFile
             }
         }
 
+        /// <summary>
+        /// Set whether to draw headpose in face Rect.
```
Comment (lebronJ) [inline app-samples/Cognitive-Services-Face-WPF/Sample-WPF/Controls/FaceDetectionPage.xaml.cs]: `Sets whether to draw face rectangles rotated by head poses`

File: app-samples/Cognitive-Services-Face-WPF/Sample-WPF/Controls/FaceDetectionPage.xaml.cs [hunk 5]
```diff
@@ -198,6 +208,42 @@ public ImageSource SelectedFile
...
+        /// </summary>
+        public bool DrawHeadPose
+        {
+            set
+            {
+                _drawHeadPose = value;
+
+                // Update face Rect here.
```
Comment (lebronJ) [inline app-samples/Cognitive-Services-Face-WPF/Sample-WPF/Controls/FaceDetectionPage.xaml.cs]: `Update face rectangles`

File: app-samples/Cognitive-Services-Face-WPF/Sample-WPF/Controls/UIHelper.cs [hunk 6]
```diff
@@ -130,13 +130,45 @@ public static IEnumerable<Face> CalculateFaceRectangleForRendering(IList<Detecte
 
             foreach (var face in faces)
             {
+                var Left = (int)((face.FaceRectangle.Left * scale) + uiXOffset);
+                var Top = (int)((face.FaceRectangle.Top * scale) + uiYOffset);
+
+                // FaceAngle use for roate face Rect, default is 0(no rotate).
```
Comment (lebronJ) [inline app-samples/Cognitive-Services-Face-WPF/Sample-WPF/Controls/UIHelper.cs]: `FaceAngle used for rotating face rectangles, default value is 0 (not rotated).`

File: app-samples/Cognitive-Services-Face-WPF/Sample-WPF/Controls/UIHelper.cs [hunk 7]
```diff
@@ -130,13 +130,45 @@ public static IEnumerable<Face> CalculateFaceRectangleForRendering(IList<Detecte
...
+                double FaceAngle = 0;
+
+                // If there has headpose attributes, will re-calculate the left/top(X/Y) position.
+                if (face.FaceAttributes?.HeadPose != null)
+                {
+                    // Headpose's roll value act directly on face angle.
```
Comment (lebronJ) [inline app-samples/Cognitive-Services-Face-WPF/Sample-WPF/Controls/UIHelper.cs]: `Headpose's roll value acts directly as the face angle.`

File: app-samples/Cognitive-Services-Face-WPF/Sample-WPF/Controls/UIHelper.cs [hunk 8]
```diff
@@ -130,13 +130,45 @@ public static IEnumerable<Face> CalculateFaceRectangleForRendering(IList<Detecte
...
+                {
+                    // Headpose's roll value act directly on face angle. 
+                    FaceAngle = face.FaceAttributes.HeadPose.Roll;
+                    var angleToPi = Math.Abs((FaceAngle / 180) * Math.PI);
+
+                    // _____       | / \ |
+                    // |____|  =>  |/   /|
+                    //             | \ / |
+                    // re-calculate the face Rect left/top(X/Y) position.
+                    var newLeft = face.FaceRectangle.Left +
```
Comment (lebronJ) [inline app-samples/Cognitive-Services-Face-WPF/Sample-WPF/Controls/UIHelper.cs:150]: rectangle

**Review by lebronJ — APPROVED** [review]
LGTM, only minor comments to change.

File: app-samples/Cognitive-Services-Face-WPF/Sample-WPF/Controls/Face.cs (hunk 1 above)
Comment (acured) [reply app-samples/Cognitive-Services-Face-WPF/Sample-WPF/Controls/Face.cs]: Done.

File: app-samples/Cognitive-Services-Face-WPF/Sample-WPF/Controls/FaceDetectionPage.xaml.cs (hunk 2 above)
Comment (acured) [reply app-samples/Cognitive-Services-Face-WPF/Sample-WPF/Controls/FaceDetectionPage.xaml.cs]: Done.

File: app-samples/Cognitive-Services-Face-WPF/Sample-WPF/Controls/FaceDetectionPage.xaml.cs (hunk 3 above)
Comment (acured) [reply app-samples/Cognitive-Services-Face-WPF/Sample-WPF/Controls/FaceDetectionPage.xaml.cs]: Done.

File: app-samples/Cognitive-Services-Face-WPF/Sample-WPF/Controls/FaceDetectionPage.xaml.cs (hunk 4 above)
Comment (acured) [reply app-samples/Cognitive-Services-Face-WPF/Sample-WPF/Controls/FaceDetectionPage.xaml.cs]: Done.

File: app-samples/Cognitive-Services-Face-WPF/Sample-WPF/Controls/FaceDetectionPage.xaml.cs (hunk 5 above)
Comment (acured) [reply app-samples/Cognitive-Services-Face-WPF/Sample-WPF/Controls/FaceDetectionPage.xaml.cs]: Done.

File: app-samples/Cognitive-Services-Face-WPF/Sample-WPF/Controls/UIHelper.cs (hunk 6 above)
Comment (acured) [reply app-samples/Cognitive-Services-Face-WPF/Sample-WPF/Controls/UIHelper.cs]: Done.

File: app-samples/Cognitive-Services-Face-WPF/Sample-WPF/Controls/UIHelper.cs (hunk 7 above)
Comment (acured) [reply app-samples/Cognitive-Services-Face-WPF/Sample-WPF/Controls/UIHelper.cs]: Done.

File: app-samples/Cognitive-Services-Face-WPF/Sample-WPF/Controls/UIHelper.cs (hunk 8 above)
Comment (acured) [reply app-samples/Cognitive-Services-Face-WPF/Sample-WPF/Controls/UIHelper.cs:150]: Done.

File: app-samples/Cognitive-Services-Face-WPF/Sample-WPF/Controls/FaceDetectionPage.xaml [hunk 9]
```diff
@@ -47,12 +47,12 @@
...
         </Grid.ColumnDefinitions>
         <Grid.RowDefinitions>
-            <RowDefinition Height="54" />
+            <RowDefinition Height="64" />
             <RowDefinition Height="45" />
             <RowDefinition Height="*" />
         </Grid.RowDefinitions>
         <TextBlock Grid.Row="0" Grid.Column="0" Grid.ColumnSpan="3" HorizontalAlignment="Left" Text="{Binding FaceDetectionDescription}" TextWrapping="Wrap"></TextBlock>
-        <Grid Width="300" Height="300" Grid.Column="0" Grid.Row="2" VerticalAlignment="Top">
+        <Grid Width="300" Height="300" Grid.Column="0" Grid.Row="3" VerticalAlignment="Top">
```
Comment (lebronJ) [inline app-samples/Cognitive-Services-Face-WPF/Sample-WPF/Controls/FaceDetectionPage.xaml]: Please double check if `Grid.Row="3"` is what you want.

File: app-samples/Cognitive-Services-Face-WPF/Sample-WPF/Controls/UIHelper.cs [hunk 10]
```diff
@@ -130,13 +130,45 @@ public static IEnumerable<Face> CalculateFaceRectangleForRendering(IList<Detecte
 
             foreach (var face in faces)
             {
+                var Left = (int)((face.FaceRectangle.Left * scale) + uiXOffset);
```
Comment (lebronJ) [inline app-samples/Cognitive-Services-Face-WPF/Sample-WPF/Controls/UIHelper.cs]: The internal `()` is unnecessary.

Comment (lebronJ) [inline app-samples/Cognitive-Services-Face-WPF/Sample-WPF/Controls/UIHelper.cs]: Change Left, Top and FaceAngle with first letter of lower case.

File: app-samples/Cognitive-Services-Face-WPF/Sample-WPF/Controls/UIHelper.cs [hunk 11]
```diff
@@ -130,13 +130,45 @@ public static IEnumerable<Face> CalculateFaceRectangleForRendering(IList<Detecte
...
+                var Top = (int)((face.FaceRectangle.Top * scale) + uiYOffset);
+
+                // FaceAngle used for rotating face rectangles, default value is 0 (not rotated).
+                double FaceAngle = 0;
+
+                // If there has headpose attributes, will re-calculate the left/top(X/Y) position.
```
Comment (lebronJ) [inline app-samples/Cognitive-Services-Face-WPF/Sample-WPF/Controls/UIHelper.cs]: `If head pose attributes have been obtained, re-calculate the left & top (X & Y) positions.`

File: app-samples/Cognitive-Services-Face-WPF/Sample-WPF/Controls/UIHelper.cs [hunk 12]
```diff
@@ -130,13 +130,45 @@ public static IEnumerable<Face> CalculateFaceRectangleForRendering(IList<Detecte
 
             foreach (var face in faces)
             {
+                var Left = (int)((face.FaceRectangle.Left * scale) + uiXOffset);
+                var Top = (int)((face.FaceRectangle.Top * scale) + uiYOffset);
+
+                // FaceAngle used for rotating face rectangles, default value is 0 (not rotated).
```
Comment (lebronJ) [inline app-samples/Cognitive-Services-Face-WPF/Sample-WPF/Controls/UIHelper.cs]: `Angle of face rectangles, default value is 0 (not rotated).`

File: app-samples/Cognitive-Services-Face-WPF/Sample-WPF/Controls/UIHelper.cs [hunk 13]
```diff
@@ -130,13 +130,45 @@ public static IEnumerable<Face> CalculateFaceRectangleForRendering(IList<Detecte
...
+                double FaceAngle = 0;
+
+                // If there has headpose attributes, will re-calculate the left/top(X/Y) position.
+                if (face.FaceAttributes?.HeadPose != null)
+                {
+                    // Headpose's roll value acts directly as the face angle.
```
Comment (lebronJ) [inline app-samples/Cognitive-Services-Face-WPF/Sample-WPF/Controls/UIHelper.cs]: `Head pose's`

File: app-samples/Cognitive-Services-Face-WPF/Sample-WPF/Controls/UIHelper.cs [hunk 14]
```diff
@@ -130,13 +130,45 @@ public static IEnumerable<Face> CalculateFaceRectangleForRendering(IList<Detecte
...
+                    var newLeft = face.FaceRectangle.Left +
+                        face.FaceRectangle.Width / 2 -
+                        (face.FaceRectangle.Width * Math.Sin(angleToPi) + face.FaceRectangle.Height * Math.Cos(angleToPi)) / 2;
+
+                    var newTop = face.FaceRectangle.Top +
+                        face.FaceRectangle.Height / 2 -
+                        (face.FaceRectangle.Height * Math.Sin(angleToPi) + face.FaceRectangle.Width * Math.Cos(angleToPi)) / 2;
+
+                    Left = (int)((newLeft * scale) + uiXOffset);
+                    Top = (int)((newTop * scale) + uiYOffset);
```
Comment (lebronJ) [inline app-samples/Cognitive-Services-Face-WPF/Sample-WPF/Controls/UIHelper.cs]: Remove the internal `()`.

**Review by lebronJ — CHANGES_REQUESTED** [review]

File: app-samples/Cognitive-Services-Face-WPF/Sample-WPF/Controls/FaceDetectionPage.xaml (hunk 9 above)
Comment (acured) [reply app-samples/Cognitive-Services-Face-WPF/Sample-WPF/Controls/FaceDetectionPage.xaml]: Revert to 2.

File: app-samples/Cognitive-Services-Face-WPF/Sample-WPF/Controls/UIHelper.cs (hunk 10 above)
Comment (acured) [reply app-samples/Cognitive-Services-Face-WPF/Sample-WPF/Controls/UIHelper.cs]: Done.

Comment (acured) [reply app-samples/Cognitive-Services-Face-WPF/Sample-WPF/Controls/UIHelper.cs]: Done.

File: app-samples/Cognitive-Services-Face-WPF/Sample-WPF/Controls/UIHelper.cs (hunk 11 above)
Comment (acured) [reply app-samples/Cognitive-Services-Face-WPF/Sample-WPF/Controls/UIHelper.cs]: Done.

File: app-samples/Cognitive-Services-Face-WPF/Sample-WPF/Controls/UIHelper.cs (hunk 12 above)
Comment (acured) [reply app-samples/Cognitive-Services-Face-WPF/Sample-WPF/Controls/UIHelper.cs]: Done.

File: app-samples/Cognitive-Services-Face-WPF/Sample-WPF/Controls/UIHelper.cs (hunk 13 above)
Comment (acured) [reply app-samples/Cognitive-Services-Face-WPF/Sample-WPF/Controls/UIHelper.cs]: Done.

File: app-samples/Cognitive-Services-Face-WPF/Sample-WPF/Controls/UIHelper.cs (hunk 14 above)
Comment (acured) [reply app-samples/Cognitive-Services-Face-WPF/Sample-WPF/Controls/UIHelper.cs]: Done.

File: app-samples/Cognitive-Services-Face-WPF/Sample-WPF/Controls/UIHelper.cs [hunk 15]
```diff
@@ -124,19 +124,51 @@ public static IEnumerable<Face> CalculateFaceRectangleForRendering(IList<Detecte
...
+                if (face.FaceAttributes?.HeadPose != null)
+                {
+                    // Head pose's roll value acts directly as the face angle.
+                    faceAngle = face.FaceAttributes.HeadPose.Roll;
+                    var angleToPi = Math.Abs((faceAngle / 180) * Math.PI);
+
+                    // _____       | / \ |
+                    // |____|  =>  |/   /|
+                    //             | \ / |
+                    // re-calculate the face rectangle left/top(X/Y) position.
```
Comment (lebronJ) [inline app-samples/Cognitive-Services-Face-WPF/Sample-WPF/Controls/UIHelper.cs]: `Re-calculate the face rectangle's left & top (X & Y) positions.`

File: app-samples/Cognitive-Services-Face-WPF/Sample-WPF/Controls/UIHelper.cs [hunk 16]
```diff
@@ -124,19 +124,51 @@ public static IEnumerable<Face> CalculateFaceRectangleForRendering(IList<Detecte
...
+                }
+
                 yield return new Face()
                 {
                     FaceId = face.FaceId?.ToString(),
-                    Left = (int)((face.FaceRectangle.Left * scale) + uiXOffset),
-                    Top = (int)((face.FaceRectangle.Top * scale) + uiYOffset),
+                    Left = left,
+                    Top = top,
+                    OriginalLeft = (int)((face.FaceRectangle.Left * scale) + uiXOffset),
```
Comment (lebronJ) [inline app-samples/Cognitive-Services-Face-WPF/Sample-WPF/Controls/UIHelper.cs]: Why not remove the internal `{}`?

File: app-samples/Cognitive-Services-Face-WPF/Sample-WPF/Controls/UIHelper.cs [hunk 17]
```diff
@@ -110,9 +110,9 @@ public static IEnumerable<Face> CalculateFaceRectangleForRendering(IList<Detecte
         {
             var imageWidth = imageInfo.Item1;
             var imageHeight = imageInfo.Item2;
-            float ratio = (float)imageWidth / imageHeight;
-            int uiWidth = 0;
-            int uiHeight = 0;
+            var ratio = (float)imageWidth / imageHeight;
+            var uiWidth = 0;
+            var uiHeight = 0;
```
Comment (lebronJ) [inline app-samples/Cognitive-Services-Face-WPF/Sample-WPF/Controls/UIHelper.cs]: [Nit] A more elegant way is to directly use
`            int uiWidth;`
`            int uiHeight;`

File: app-samples/Cognitive-Services-Face-WPF/Sample-WPF/Controls/UIHelper.cs (hunk 16 above)
Comment (lebronJ) [reply app-samples/Cognitive-Services-Face-WPF/Sample-WPF/Controls/UIHelper.cs]: Same for the below.

**Review by lebronJ — CHANGES_REQUESTED** [review]
Final comments.

File: app-samples/Cognitive-Services-Face-WPF/Sample-WPF/Controls/UIHelper.cs (hunk 15 above)
Comment (acured) [reply app-samples/Cognitive-Services-Face-WPF/Sample-WPF/Controls/UIHelper.cs]: Done.

File: app-samples/Cognitive-Services-Face-WPF/Sample-WPF/Controls/UIHelper.cs (hunk 16 above)
Comment (acured) [reply app-samples/Cognitive-Services-Face-WPF/Sample-WPF/Controls/UIHelper.cs]: Done.

File: app-samples/Cognitive-Services-Face-WPF/Sample-WPF/Controls/UIHelper.cs (hunk 17 above)
Comment (acured) [reply app-samples/Cognitive-Services-Face-WPF/Sample-WPF/Controls/UIHelper.cs]: Done.

**Review by lebronJ — APPROVED** [review]

**Pull request closed without merging**